ads/google/csa.js

```diff
@@ -0,0 +1,332 @@
+  }
+}
+
+/**
+ * Hanlder for When a resize request is denied
+ * If the container is larger than the AMP container and an overflow already
+ * exists, show the overflow and resize the container to fit inside the AMP
+ * container.  If an overflow doesn't exist, create one.
+ * @visibleForTesting
+ */
+export function resizeDeniedHandler(global, container, requestedHeight) {
+  const overflow = global.document.getElementById('overflow');
+  const containerHeight = parseInt(container.style.height, 10);
+  if (containerHeight > currentAmpHeight) {
+    if (overflow) {
+      overflow.style.display = '';
+      resizeCsa(container, currentAmpHeight - overflowHeight);
+    } else {
+      createOverflow(global, container, requestedHeight);
+    }
+  }
+}
+
+/**
+ * Make a request for either AFS or AFSh
+ * @param {!Window} global The window object of the iframe
+ * @param {!Object} data The data passed in by the partner
+ * @param {!Object} afsP The parsed AFS page options object
+ * @param {!Object} afsA The parsed AFS adblock options object
+ * @param {!Object} afshP The parsed AFSh page options object
+ * @param {!Object} afshA The parsed AFSh adblock options object
+ */
+function requestCsaAds(global, data, afsP, afsA, afshP, afshA) {
+  const type = getAdType(data);
+
+  switch (type) {
+    case AD_TYPE.AFS:
+      /** Do not backfill, request AFS */
+      afsA['adLoadedCallback'] = resizeIframe.bind(null, global, null, null);
+      global._googCsa('ads', afsP, afsA);
+      break;
+    case AD_TYPE.AFSH:
+      /** Do not backfill, request AFSh */
+      afshA['adLoadedCallback'] = resizeIframe.bind(null, global, null, null);
+      global._googCsa('plas', afshP, afshA);
+      break;
+    case AD_TYPE.AFSH_BACKFILL:
+      /** Backfill with AFS, request AFSh */
+      afshA['adLoadedCallback'] = resizeIframe.bind(null, global, afsP, afsA);
+      global._googCsa('plas', afshP, afshA);
+      break;
+    default:
+      return;
+  }
+}
+
+/**
+ * Helper function to determine which product to request
+ * @param {!Object} data The data passed in by the partner
+ * @return {number} Enum of ad type
+ */
+function getAdType(data) {
+  if (data['afsPageOptions'] != null && data['afshPageOptions'] == null) {
+    return AD_TYPE.AFS;
+  }
+  if (data['afsPageOptions'] == null && data['afshPageOptions'] != null) {
+    return AD_TYPE.AFSH;
+  }
+  if (data['afsPageOptions'] != null && data['afshPageOptions'] != null) {
+    return AD_TYPE.AFSH_BACKFILL;
+  }
+  else {
+    return AD_TYPE.UNSUPPORTED;
+  }
+}
+
+/**
+ * CSA callback function to resize the iframe and/or request backfill
+ * @param {?Object} bkflPage Backfill AFS page options (if necessary)
+ * @param {?Object} bkflAd Backfill AFS ad unit options (if necessary)
+ * @param {string} conName Name of the container ('csacontainer')
+ * @param {boolean} ads If ads were returned or not
+ * @visibleForTesting
+ */
+export function resizeIframe(global, bkflPage, bkflAd, conName, ads) {
```

Comment:
shall we just split this function into 2?
```
function resizeIframe(global, containerName, hasAd) {}
function backfill(global, backfillPage, backfillAd) {}
```

it gives not just shorter argument list, but also avoided the recursive call of resizeIframe.

Then the above listener code becomes:
```
afshA['adLoadedCallback'] = (containerName, hasAd) => {
  if (hasAd) {
    resizeIframe(global, containerName, hasAd);
  } else {
     backfill(global, afsP, afsA)
  }
} 
```